File: Week 28/game_theory.py
# Week 28: Game Theory & Combinatorics
# Nim, Sprague-Grundy, Pascal's Triangle, Catalan, Matrix Exponentiation

# === Nim Game ===
# Time: O(n), Space: O(1)
def nim_winner(piles):
    xor = 0
    for p in piles:
        xor ^= p
    return "First" if xor != 0 else "Second"

# === Sprague-Grundy ===
# Time: O(n * |moves|), Space: O(n)
def grundy(n, moves):
    g = [0] * (n + 1)
    for i in range(1, n + 1):
        reachable = set()
        for m in moves:
            if i >= m:
                reachable.add(g[i - m])
        mex = 0
        while mex in reachable:
            mex += 1
        g[i] = mex
    return g[n]

# === Pascal's Triangle / nCr ===
# Time: O(n^2), Space: O(n^2)
def build_pascal(n):
    C = [[0] * (n + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        C[i][0] = 1
        for j in range(1, i + 1):
            C[i][j] = C[i-1][j-1] + C[i-1][j]
    return C

# === Catalan Numbers ===
# Time: O(n), Space: O(1) using binomial formula
def catalan(n):
    if n <= 1:
        return 1
    result = 1
    for i in range(n):
        result = result * (2 * n - i) // (i + 1)
    return result // (n + 1)

# === Matrix Exponentiation for Fibonacci ===
# Time: O(k^3 * log n), Space: O(k^2)
def mat_mult(A, B, mod=10**9+7):
    n = len(A)
    C = [[0]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % mod
    return C

def mat_pow(M, p, mod=10**9+7):
    n = len(M)
    result = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
    while p > 0:
        if p & 1:
            result = mat_mult(result, M, mod)
        M = mat_mult(M, M, mod)
        p >>= 1
    return result

def fibonacci(n, mod=10**9+7):
    if n <= 1:
        return n
    M = [[1, 1], [1, 0]]
    result = mat_pow(M, n - 1, mod)
    return result[0][0]

if __name__ == "__main__":
    print(f"Nim [3,4,5]: {nim_winner([3,4,5])}")
    print(f"Grundy(10, moves=[1,3,4]): {grundy(10, [1,3,4])}")
    C = build_pascal(10)
    print(f"C(10,3) = {C[10][3]}")
    print(f"Catalan(5) = {catalan(5)}")
    print(f"Fibonacci(10) = {fibonacci(10)}")

File: Week 28/java/game_theory.java
// Week 28: Game Theory & Combinatorics
// Nim O(n), Sprague-Grundy, Pascal's Triangle, Catalan, Matrix Exponentiation

import java.util.*;

public class game_theory {

    // Nim Game - XOR all pile sizes
    static String nimWinner(int[] piles) {
        int xor = 0;
        for (int p : piles) xor ^= p;
        return xor != 0 ? "First" : "Second";
    }

    // Sprague-Grundy - compute Grundy number
    static int grundy(int n, int[] moves) {
        int[] g = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            Set<Integer> reachable = new HashSet<>();
            for (int m : moves)
                if (i >= m) reachable.add(g[i - m]);
            int mex = 0;
            while (reachable.contains(mex)) mex++;
            g[i] = mex;
        }
        return g[n];
    }

    // Pascal's Triangle
    static long[][] buildPascal(int n) {
        long[][] C = new long[n+1][n+1];
        for (int i = 0; i <= n; i++) {
            C[i][0] = 1;
            for (int j = 1; j <= i; j++)
                C[i][j] = C[i-1][j-1] + C[i-1][j];
        }
        return C;
    }

    // Catalan Number using binomial formula
    static long catalan(int n) {
        long result = 1;
        for (int i = 0; i < n; i++)
            result = result * (2 * n - i) / (i + 1);
        return result / (n + 1);
    }

    // Matrix Exponentiation for Fibonacci
    static long MOD = 1_000_000_007;

    static long[][] matMult(long[][] A, long[][] B) {
        int n = A.length;
        long[][] C = new long[n][n];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                for (int k = 0; k < n; k++)
                    C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD;
        return C;
    }

    static long[][] matPow(long[][] M, long p) {
        int n = M.length;
        long[][] result = new long[n][n];
        for (int i = 0; i < n; i++) result[i][i] = 1;
        while (p > 0) {
            if ((p & 1) == 1) result = matMult(result, M);
            M = matMult(M, M);
            p >>= 1;
        }
        return result;
    }

    static long fibonacci(int n) {
        if (n <= 1) return n;
        long[][] M = {{1, 1}, {1, 0}};
        return matPow(M, n - 1)[0][0];
    }

    public static void main(String[] args) {
        System.out.println("Nim [3,4,5]: " + nimWinner(new int[]{3,4,5}));
        System.out.println("Grundy(10, [1,3,4]): " + grundy(10, new int[]{1,3,4}));
        long[][] C = buildPascal(10);
        System.out.println("C(10,3) = " + C[10][3]);
        System.out.println("Catalan(5) = " + catalan(5));
        System.out.println("Fibonacci(10) = " + fibonacci(10));
    }
}
